File: api_connect.js
const fs = require('fs');
const { url } = require('inspector');
const file_saver = require('./file_saver.js');
const { get } = require('http');
const os = require("os");
const { dir } = require('console');
const check_files = require('./check_files.js');
const print = require("./extras.js")


let code = undefined;
let token = undefined;
let refresh_token = undefined;

const userHomeDir = os.homedir()
const dir_path = userHomeDir + "/cloud_storage";


async function call_api_graph(input)
{
    const url = "https://graph.microsoft.com/v1.0" + input;

    return await fetch(url, {
        method: 'GET',
        headers: {
            'Authorization': `Bearer ${token}`
        }
    })
        .then(response => response.json())
        .then(data =>
        {
            return data;
        })
        .catch((error) =>
        {
            console.error('Error:', error);
            return -1;
        });
}


async function get_new_token_with_refresh_token(settings)
{
    console.log("\n\nREFRESHING TOKEN\n\n")
    const host = "https://login.microsoftonline.com";
    const tenantId = settings.tenantId;
    const clientId = settings.clientId;
    const clientSecret = settings.clientSecret;
    const scope = "https://graph.microsoft.com/.default";
    const grantType = "refresh_token";

    const url = `${host}/${tenantId}/oauth2/v2.0/token`;
    const body = `client_id=${clientId}&scope=${scope}&refresh_token=${refresh_token}&grant_type=${grantType}&client_secret=${clientSecret}`;

    const response = await fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded'
        },
        body: body
    });

    const data = await response.json();
    token = data.access_token;
    refresh_token = data.refresh_token;

    let res = await call_api_graph('/me');
    if (res.error !== undefined)
    {
        console.log("Error: " + res.error.message);
        return res.error;
    } else
    {
        file_saver.save_tokens_on_file(token, refresh_token);
        console.log("Logged in");
        console.log(res);
        return 0;
    }
}


async function api_login(settings)
{
    let file_tokens = file_saver.get_tokens_from_file();

    if (file_tokens !== undefined)
    {
        token = file_tokens.token;
        refresh_token = file_tokens.refresh_token;
        let res = await call_api_graph('/me');
        if (res.error !== undefined)
        {
            console.log("Error: " + res.error.message);
        } else
        {
            console.log("Logged in");
            console.log(res);
            return 0;
        }
    }


    if (await get_new_token_with_refresh_token(settings) === 0)
    {
        return 0;
    }

    let url_fetch = file_saver.construct_url(settings);
    let response = await fetch(url_fetch, {
        method: 'GET',
    });
    console.log("\n\n\nOPEN THIS URL IN YOUR BROWSER: " + response.url)
    return 0;
}

async function get_tokens(settings) //ask login on link
{
    const host = "https://login.microsoftonline.com";
    const tenantId = settings.tenantId;
    const clientId = settings.clientId;
    const clientSecret = settings.clientSecret;
    const scope = "https://graph.microsoft.com/.default";
    const grantType = "authorization_code";

    const url = `${host}/${tenantId}/oauth2/v2.0/token`;
    const body = `client_id=${clientId}&scope=${scope}&code=${code}&redirect_uri=http://localhost:8080&grant_type=${grantType}&client_secret=${clientSecret}`;

    const response = await fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded'
        },
        body: body
    });

    const data = await response.json();
    token = data.access_token;
    refresh_token = data.refresh_token;

    file_saver.save_tokens_on_file(token, refresh_token);

    return data;
}



function upload_file(file_path, file_name)
{
    const url = "https://graph.microsoft.com/v1.0/me/drive/root:/cloud_storage" + file_name + ":/content";

    fetch(url, {
        method: 'PUT',
        headers: {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json'
        },
        body: fs.readFileSync(file_path)
    })
        .then(response => response.json())
        .then(data =>
        {
            console.log(data);
        })
        .catch((error) =>
        {
            console.error('Error:', error);
        });

    return 0;
}



async function download_file(file_name)
{
    const url = "https://graph.microsoft.com/v1.0/me/drive/root:/cloud_storage" + file_name + ":/content";

    let res = await fetch(url, {
        method: 'GET',
        headers: {
            'Authorization': `Bearer ${token}`,
        }
    })
        .then(data =>
        {
            return data;
        })
        .catch((error) =>
        {
            console.error('Error:', error);
            return -1;
        });

    if (res === -1)
        return -1;


    let data = await res.blob();
    let buffer = await data.arrayBuffer();

    //create dir if does not exist
    let split_path = file_name.split("/");

    split_path.pop();
    let temp_path = dir_path;

    for (let i = 0; i < split_path.length; i++)
    {
        if (split_path[i] == "")
            continue;
        temp_path += "/" + split_path[i];
        if (!fs.existsSync(temp_path))
        {
            fs.mkdirSync(temp_path);
        }
    }

    file_name = file_name.replace(dir_path, "");

    try
    {
        let file_buffer_on_pc = fs.readFileSync(dir_path + file_name);

        if (check_files.is_file_changed(file_buffer_on_pc, Buffer.from(buffer)))
        {
            fs.writeFileSync(dir_path + file_name, Buffer.from(buffer));
            console.log("File changed, downloaded at: " + dir_path + file_name + " size of file: " + buffer.byteLength);
        }

    } catch (err)
    {
        if (err.code === 'ENOENT')
        {
            fs.writeFileSync(dir_path + file_name, Buffer.from(buffer));
            console.log("File downloaded at: " + dir_path + file_name + " size of file: " + buffer.byteLength);
        }
    }
    return 0;
}


async function get_all_files(path_to_folder)
{
    const path = "/me/drive/root:/cloud_storage" + path_to_folder + ":/children";

    let res = await fetch("https://graph.microsoft.com/v1.0" + path, {
        method: 'GET',
        headers: {
            'Authorization': `Bearer ${token}`
        }
    })
        .then(response => response.json())
        .then(data =>
        {
            return data;
        })
        .catch((error) =>
        {
            console.error('Error:', error);
            return -1;
        });

    //if res contains folder print folder
    if (res === -1)
        return -1;
    let files = res.value;
    for (let i = 0; i < files.length; i++)
    {
        if (files[i].folder != undefined)
        {
            console.log("Folder Found: " + files[i].name);
            get_all_files(path_to_folder + "/" + files[i].name);
        } else
        {
            let final_path = path_to_folder + "/" + files[i].name
            download_file(final_path);
        }

    }
}


module.exports = {
    api_login,
    call_api_graph,
    get_tokens,
    upload_file,
    download_file,
    get_all_files,
};

File: api_connect.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/go-zoox/fetch"
)

type Tokens struct {
	Token        string `json:"token"`
	RefreshToken string `json:"refresh_token"`
}

var global_code string = ""
var global_tokens Tokens = Tokens{}

func fetch_upload_file(file *os.File, url string) (*fetch.Response, error) {

	testprint := strings.Replace(url, "https://graph.microsoft.com/v1.0/me/drive/root:/cloud_storage/", "", 1)
	fmt.Println("Uploading file:", strings.Replace(testprint, ":/content", "", 1))
	response, err := fetch.Put(url, &fetch.Config{
		Headers: map[string]string{
			"Authorization": "Bearer " + global_tokens.Token,
			"Content-Type":  "application/octet-stream",
		},
		Body: file,
	})
	return response, err
}

func upload_file_to_onedrive(file_path string, file_name_on_cloud string) {
	defer wg.Done()
	defer func() { number_of_files-- }()

	url := "https://graph.microsoft.com/v1.0/me/drive/root:/cloud_storage/" + file_name_on_cloud + ":/content"

	file, err := os.Open(file_path)
	if err != nil {
		fail("Error opening file:", err)
		return
	}
	defer file.Close()

	response, err := fetch_upload_file(file, url)

	if err != nil {
		fail("Error uploading file:", err)
		number_of_errors++
		return
	}

	fmt.Println("File uploaded with code:", response.StatusCode())
}

func delete_file_on_onedrive(file_name_on_cloud string) {
	url := "https://graph.microsoft.com/v1.0/me/drive/root:/cloud_storage/" + file_name_on_cloud

	response, err := fetch.Delete(url, &fetch.Config{
		Headers: map[string]string{
			"Authorization": "Bearer " + global_tokens.Token,
		},
	})

	if err != nil {
		fail("Error deleting file:", err)
		return
	}

	fmt.Println("File deleted with code:", response.StatusCode())
}

func donwload_file_from_onedrive(file_name_on_cloud string) {
	url := "https://graph.microsoft.com/v1.0/me/drive/root:/cloud_storage/" + file_name_on_cloud + ":/content"

	response, err := fetch.Get(url, &fetch.Config{
		Headers: map[string]string{
			"Authorization": "Bearer " + global_tokens.Token,
		},
	})

	if err != nil {
		fail("Error downloading file:", err)
		return
	}

	file_path := "/home/marques/cloud_storage/" + file_name_on_cloud

	file, err := os.Create(file_path)
	if err != nil {
		fail("Error creating file:", err)
		return
	}
	defer file.Close()

	_, err = file.Write(response.Body)
	if err != nil {
		fail("Error writing file:", err)
		return
	}

	fmt.Println("File downloaded with code:", response.StatusCode())
}

// api login functions
func call_graph_api(call string, tokens Tokens) (*fetch.Response, error) {
	url := "https://graph.microsoft.com/v1.0/" + call

	response, err := fetch.Get(url, &fetch.Config{
		Headers: map[string]string{
			"Authorization": "Bearer " + tokens.Token,
		},
	})

	if err != nil {
		fmt.Println("Error fetching data:", err)
		return nil, err
	}

	return response, nil
}

func get_tokens_from_file() (Tokens, error) {

	var tokens Tokens

	tokens_read, err := os.ReadFile("tokens.json")
	if err != nil {
		return tokens, err
	}

	err = json.Unmarshal(tokens_read, &tokens)
	if err != nil {
		fmt.Println("Error unmarshaling JSON:", err)
		return tokens, err
	}

	return tokens, nil
}

func create_login_link(settings Settings) {
	host := "https://login.microsoftonline.com"
	path := settings.TenentID + "/oauth2/v2.0/authorize"

	url := host + "/" + path

	url += "?client_id=" + settings.ClientID
	url += "&response_type=code"
	url += "&redirect_uri=http://localhost:8080"
	url += "&response_mode=query"
	url += "&scope=" + settings.GraphUserScope
	url += "&state=12345"

	global_code = ""
	fmt.Println("Login link:", url)

	for global_code == "" {
		time.Sleep(1 * time.Second)
	}

	fmt.Println("Code received:", global_code)
}

func refresh_token(tokens Tokens, settings Settings, is_code bool) (Tokens, error) {
	fmt.Println("Refreshing token")

	host := "https://login.microsoftonline.com"
	path := settings.TenentID + "/oauth2/v2.0/token"

	url := host + "/" + path

	var body map[string]string
	if is_code {
		body = map[string]string{
			"client_id":     settings.ClientID,
			"scope":         "https://graph.microsoft.com/.default offline_access",
			"code":          global_code,
			"redirect_uri":  "http://localhost:8080",
			"grant_type":    "authorization_code",
			"client_secret": settings.ClientSecret,
		}
	} else {
		body = map[string]string{
			"client_id":     settings.ClientID,
			"scope":         "https://graph.microsoft.com/.default offline_access",
			"refresh_token": tokens.RefreshToken,
			"grant_type":    "refresh_token",
			"client_secret": settings.ClientSecret,
		}
	}

	response, err := fetch.Post(url, &fetch.Config{
		Headers: map[string]string{
			"Content-Type": "application/x-www-form-urlencoded",
		},
		Body: body,
	})

	if err != nil {
		fmt.Println("Error fetching data:", err)
		return tokens, err
	}

	if response.StatusCode() != 200 {
		fmt.Println("Error refreshing token, status code:", response.StatusCode())
		return tokens, fmt.Errorf("error refreshing token, status code: %d", response.StatusCode())
	}

	json_string, err := response.JSON()
	if err != nil {
		fmt.Println("Error parsing JSON:", err)
		return tokens, err
	}

	//convert json string to struct

	new_tokens := struct {
		Token        string `json:"access_token"`
		RefreshToken string `json:"refresh_token"`
		Scope        string `json:"scope"`
		TokenType    string `json:"token_type"`
	}{}
	err = json.Unmarshal([]byte(json_string), &new_tokens)
	if err != nil {
		fmt.Println("Error unmarshaling JSON:", err)
		return tokens, err
	}

	tokens.Token = new_tokens.Token
	tokens.RefreshToken = new_tokens.RefreshToken

	return tokens, nil
}

func login_the_account(tokens Tokens, settings Settings) (Tokens, error) {
	res, err := call_graph_api("/me", tokens)
	if err != nil {
		fmt.Println("Error calling graph API:", err)
		return tokens, err
	}

	status_code := res.StatusCode()

	if status_code == 401 {
		fmt.Println("Token expired, refreshing token")
		new_tokens, err := refresh_token(tokens, settings, false)
		if err != nil {
			fmt.Println("Error refreshing token:", err)

			create_login_link(settings)
			new_tokens, err = refresh_token(tokens, settings, true)

			if err != nil {
				fmt.Println("Error refreshing token:", err)
				return tokens, err
			}
		}
		tokens = new_tokens
	}

	if status_code != 200 && status_code != 401 {
		fmt.Println("Error calling graph API, status code:", status_code)
		return tokens, fmt.Errorf("error calling graph API, status code: %d", status_code)
	}

	res, err = call_graph_api("/me", tokens)
	if err != nil || res.StatusCode() != 200 {
		fmt.Println("Cannot login you:", err)
		return tokens, err
	}
	okay("Logged in successfully")

	return tokens, nil
}

func save_tokens_to_file(tokens Tokens) error {
	tokens_json, err := json.Marshal(tokens)
	if err != nil {
		fmt.Println("Error marshaling JSON:", err)
		return err
	}

	err = os.WriteFile("tokens.json", tokens_json, 0644)
	if err != nil {
		fmt.Println("Error writing file:", err)
		return err
	}

	return nil
}

func start_api(settings Settings) {
	info("Starting API")

	tokens, err := get_tokens_from_file()
	if err != nil {
		fmt.Println("Error getting tokens from file:", err)
		return
	}

	new_tokens, nil := login_the_account(tokens, settings)

	if err != nil {
		fmt.Println("Error logging in:", err)
		return
	}

	if tokens.Token != new_tokens.Token {
		fmt.Println("Token refreshed, saving to file")
		tokens = new_tokens
		err = save_tokens_to_file(tokens)
		if err != nil {
			fmt.Println("Error saving tokens to file:", err)
			return
		}
	}

	global_tokens = tokens
	okay("Tokens set, it`s ready")
}
